Typing slice: annotate src/bernstein/tui/ (part of the mypy backlog)

Part of the typing cleanup tracked in #2980. That issue covers ~1017 mypy errors across 312 files, which is far too large for one pull request, so it is being cut into small independent slices. **This slice is one of them: the `src/bernstein/tui/` package only.**

No prior knowledge of the codebase is required. Everything you need is below.

## Why this matters

`tui` is the terminal dashboard an operator watches during a live run — task list, approval panel, notifications, command palette. A type slip here means a UI element quietly renders the wrong value, or the app crashes mid-session, instead of the type checker catching the mismatch before anyone runs it.

## Setup

```bash
git clone https://github.com/sipyourdrink-ltd/bernstein
cd bernstein
uv sync
```

## Step 1 — see the errors you are fixing

```bash
uv run mypy src/bernstein/tui/ 2>&1 | tail -30
```

That prints the current error list plus a count. **Copy that count somewhere** — you will compare against it at the end. At the time this issue was written, that command reported **41 errors in 7 files**. Your number may be slightly different if the codebase has moved since — that's fine, just use whatever Step 1 gives you as your own baseline.

## Step 2 — fix them

Work file by file. There are two recurring shapes, plus one cluster of errors that isn't yours to fix in this slice (see "What to skip").

**a) A module-level "type alias" that's actually just a string**

`app.py` defines a module-level constant that holds a type *as a string*, then passes it to `cast()` repeatedly:

```python
# before — src/bernstein/tui/app.py
# Shared cast-type constants to avoid string duplication (Sonar S1192).
_CAST_DICT_STR_ANY = "dict[str, Any]"

...

tasks_payload = cast(_CAST_DICT_STR_ANY, tasks_section)
```

A plain `str` variable isn't a type as far as mypy is concerned, so it reports `Variable "..." is not valid as a type` at every `cast()` call, and then every attribute access on the "cast" result fails too, because the cast never actually happened from mypy's point of view. Fix it by turning the constant into a real type alias with Python 3.12's `type` statement — none of the `cast()` call sites (there are about ten of them in this file) need to change:

```python
# after
type _CAST_DICT_STR_ANY = dict[str, Any]
```

**b) `float()`/`int()` called directly on a value pulled out of a `dict[str, object]`**

`approval_panel.py` renders an SLO burndown widget from a loosely-typed data dict:

```python
# before — src/bernstein/tui/approval_panel.py
def build_slo_burndown_text(burndown: dict[str, object]) -> Text:
    ...
    slo_target = float(burndown.get("slo_target", 0.9))
    slo_current = float(burndown.get("slo_current", 0.0))
    burn_rate = float(burndown.get("burn_rate", 0.0))
    budget_fraction = float(burndown.get("budget_fraction", 1.0))
    budget_consumed_pct = float(burndown.get("budget_consumed_pct", 0.0))
```

`burndown` is declared `dict[str, object]`, so `.get(...)` always returns `object`, and `float()` doesn't accept a bare `object`. The fix is to tell mypy what shape you actually expect back:

```python
# after
    slo_target = float(cast(float, burndown.get("slo_target", 0.9)))
    slo_current = float(cast(float, burndown.get("slo_current", 0.0)))
    burn_rate = float(cast(float, burndown.get("burn_rate", 0.0)))
    budget_fraction = float(cast(float, burndown.get("budget_fraction", 1.0)))
    budget_consumed_pct = float(cast(float, burndown.get("budget_consumed_pct", 0.0)))
```

You'll need to add `cast` to the existing `from typing import Any, ClassVar` line. There's an `int(burndown.get("total_tasks", 0))` a line or two below this that already carries a `# type: ignore[arg-type]` — the same `cast()` approach (`int(cast(int, burndown.get("total_tasks", 0)))`) fixes it properly, and then you can delete that ignore comment.

**What to skip**

`app.py` has errors like `"Notifications" has no attribute "mark_all_read"` / `"get_history"` / `"get_unread_count"`. Unlike the `_CAST_` errors above, these look like the `Notifications` class may genuinely be missing those methods rather than mypy failing to see something that exists — that's a behaviour question, not a typing one. Leave these and mention them in your PR rather than guessing at an implementation.

Rules for this slice:

- **Do not change any runtime behaviour.** Annotations only. If fixing a type properly would require changing what the code *does*, skip that one and mention it in the PR description instead.
- **Do not add `# type: ignore`.** The point is to remove errors, not silence them. If you hit one you genuinely cannot type, leave it and say so in the PR.
- Prefer modern syntax: `str | None` over `Optional[str]`, `list[str]` over `List[str]`, and the `type X = Y` statement over `TypeAlias` for new aliases.
- Import types you need from where the codebase already imports them — grep for an existing import of the same name rather than inventing a path.

## Step 3 — verify

```bash
uv run mypy src/bernstein/tui/ 2>&1 | tail -5     # error count must be lower than Step 1
uv run ruff check src/bernstein/tui/
uv run ruff format --check src/bernstein/tui/
uv run pytest tests/unit/test_slo_burndown.py tests/unit/test_tui.py -q
```

⚠️ **Do not run the whole test suite** (`uv run pytest` with no arguments). It is very memory-hungry and will likely hang or crash your machine. Run only the specific test files above.

All four commands must pass, and the mypy count must be strictly lower than what you recorded in Step 1.

## Step 4 — open the pull request

```bash
git checkout -b typing/tui-slice
git add -A
git commit -m "chore(typing): annotate the tui package"
git push -u origin typing/tui-slice
```

In the PR description, paste the before/after mypy counts from Steps 1 and 3, and list anything you deliberately skipped (at minimum, the `Notifications` attribute errors above).

## Done means

- [ ] `uv run mypy src/bernstein/tui/` reports fewer errors than before, with the before/after numbers in the PR body
- [ ] No `# type: ignore` added
- [ ] No runtime behaviour changed — annotations only
- [ ] `ruff check` and `ruff format --check` clean on the touched files
- [ ] The two named test files pass

## If you get stuck

Comment here with the exact command you ran and its full output. A partial pull request that fixes some files correctly is more useful than a complete one that changes behaviour, so do not feel you must clear every error in the package — the `Notifications` errors in particular are fine to leave for someone else.

Assigned to you, @ShlokShar — thanks for picking it up.

Before you start: **do not follow Step 2 of the issue body verbatim.** It is nineteen days stale and most of what it describes is already fixed. I re-ran the baseline against current `main` (`c763b43`) just now so you are working from real numbers rather than the ones in the body:

```
$ uv run mypy src/bernstein/tui/
Found 9 errors in 2 files (checked 46 source files)
```

Not 41. Step 2a (the `_CAST_DICT_STR_ANY` string-alias cluster) and Step 2b (the `float()`/`int()` casts in `approval_panel.py`) both landed already, in #3236 and follow-ups. If you clone and go looking for them you will not find them, and that is the issue's fault, not yours.

### What is actually left

Of the 9 remaining errors, 8 are the `Notifications` cluster in `app.py` (`:284`, `:703`, `:1118` ×3, `:1119`, `:1130` ×2). Those stay **out of scope**, exactly as the body says — they are a real API mismatch against `textual.notifications`, not an annotation gap, and resolving them means deciding what that API should be. There is no open issue for that cluster yet; I have flagged it for the operator rather than leaving it for you to guess at.

That leaves **one** error, and it is the whole of your slice:

```
src/bernstein/tui/task_list.py:426: error: Argument 1 to "set" has incompatible type
  "dict[RowKey, Row]"; expected "Iterable[str]"  [arg-type]
```

I am not going to pretend that is a big task. It is one line. Take it as a first patch to get the contribution loop working end to end, and if you want something meatier afterwards say so and I will point you at one I have verified is still current — I would rather do that than have you find out at the end of an evening that a second issue had also gone stale.

### One correction to the record, because it changes what you should do

An earlier comment of mine in this thread lists `task_list.py:426` in a table of "likely pre-existing bugs, not typing problems", with the symptom *"stale-row cleanup probably never matches"*, and says not to touch such rows with `cast()`. **That was wrong, and you should disregard it for this line.** I checked it rather than repeating it:

```
$ python -c "from textual.widgets.data_table import RowKey; k=RowKey('abc'); print(k=='abc', hash(k)==hash('abc'), {k}-{'abc'})"
True True set()
```

On textual 8.2.1 `RowKey` subclasses `StringKey`, which defines `__eq__` and `__hash__` against the underlying string. So `existing_keys - incoming.keys()` at `:429` and `row.task_id in existing_keys` at `:445` both match correctly at runtime. The cleanup works. What is wrong is only the annotation: the variable is declared `set[str]` and actually holds `set[RowKey]`.

That matters because it puts the line squarely back **in** scope for this slice — it is an annotation-only fix, no runtime behaviour changes, which is exactly what the rules here allow.

### The decision I am leaving you

There are two honest fixes and I verified both typecheck clean on `c763b43`. Pick one and say why in the PR body; I will back either.

**A — name the type the container actually holds.**
```python
from textual.widgets.data_table import RowKey   # new import
...
existing_keys: set[RowKey] = set(self.rows)
```

**B — normalise at the boundary so everything downstream stays `str`.**
```python
existing_keys: set[str] = {str(key.value) for key in self.rows}
```

The trade-off is real in both directions. A is smaller and describes the object truthfully, but it leaves the `RowKey`/`str` interop implicit — the code silently depends on `StringKey.__eq__` continuing to compare against bare strings, and nothing in the file says so. B makes that dependency explicit and disappears if Textual ever changes the key type, but it builds a fresh set on every `refresh_tasks` call and reaches into `.value`. This is a genuine judgement call about where you want the coupling to live, not a trick question.

One trap if you take B: `str(key)` gives you the object repr, not the key — you need `key.value`.
```
value: 't-1' | str(k): '<textual.widgets._data_table.RowKey object at 0x...>'
```

### The test that proves it

`refresh_tasks` currently has only a smoke test — `tests/unit/test_tui.py:178` instantiates the widget and asserts it is not `None`. Nothing exercises the stale-row path at all, which is why an annotation could sit there disagreeing with reality for this long.

The test worth adding drives `refresh_tasks` twice: once with two rows, then again with only the first, and asserts the second row is actually gone and the first survived. That pins the behaviour the annotation is describing, and it fails loudly if someone later "fixes" the type by changing what the code does instead of what it claims. Put it alongside the existing widget tests in `tests/unit/test_tui.py`.

### Verifying

The commands in Step 3 are still correct. Your bar is **9 → 8**, not the 41 in the body. Please do not run bare `pytest` — scope it to the named files.

Anything unclear or anything that does not reproduce, comment here with the command and its full output and I will pick it up.


---

--- a/src/bernstein/tui/task_list.py
+++ b/src/bernstein/tui/task_list.py
@@ -9,6 +9,7 @@ from typing import Any
 
 from rich.text import Text
 from textual.widgets import DataTable, Static
+from textual.widgets.data_table import RowKey
 
 from bernstein.tui.accessibility import accessible_status_label, replace_unicode
 
@@ -423,7 +424,7 @@ class TaskListWidget(DataTable[Text]):
         """
         # Build a lookup of incoming rows by task_id
         incoming: dict[str, TaskRow] = {r.task_id: r for r in rows}
-        existing_keys: set[str] = set(self.rows)
+        existing_keys: set[RowKey] = set(self.rows)
 
         # Remove rows no longer present
         for key in existing_keys - incoming.keys():

--- a/src/bernstein/tui/task_list.py
+++ b/src/bernstein/tui/task_list.py
@@ -5,13 +5,15 @@ from __future__ import annotations
 import contextlib
 import time
 from dataclasses import dataclass
-from typing import Any
+from typing import TYPE_CHECKING, Any
 
 from rich.text import Text
 from textual.widgets import DataTable, Static
-from textual.widgets.data_table import RowKey
 
 from bernstein.tui.accessibility import accessible_status_label, replace_unicode
+
+if TYPE_CHECKING:
+    from textual.widgets.data_table import RowKey
 
 # Sparkline characters for cost trend visualization
 SPARKLINE_CHARS = "▁▂▃▄▅▆▇█"

--- a/tests/unit/test_tui.py
+++ b/tests/unit/test_tui.py
@@ -192,6 +192,39 @@ class TestWidgetCreation:
         assert widget is not None
 
 
+class TestTaskListRefresh:
+    """Tests for stale-row cleanup in TaskListWidget."""
+
+    def test_refresh_tasks_removes_stale_rows(self) -> None:
+        widget = TaskListWidget()
+        widget.on_mount()
+        first = TaskRow(
+            task_id="t-1",
+            status="open",
+            role="worker",
+            title="First task",
+            priority=2,
+            model="test",
+            elapsed="1s",
+            session_id="s1",
+        )
+        second = TaskRow(
+            task_id="t-2",
+            status="open",
+            role="worker",
+            title="Second task",
+            priority=2,
+            model="test",
+            elapsed="1s",
+            session_id="s2",
+        )
+        widget.refresh_tasks([first, second])
+        assert {str(key.value) for key in widget.rows} == {"t-1", "t-2"}
+
+        widget.refresh_tasks([first])
+        assert [str(key.value) for key in widget.rows] == ["t-1"]
+
+
 class TestAppInstantiation:
     """Tests that the Textual app can be created."""
 

--- a/tests/unit/test_tui.py
+++ b/tests/unit/test_tui.py
@@ -5,6 +5,7 @@ from __future__ import annotations
 from typing import TYPE_CHECKING, Any
 
 import pytest
+from textual.app import App
 
 from bernstein.tui.app import BernsteinApp, _kill_agent, _kill_all_agents
 from bernstein.tui.widgets import (
@@ -195,9 +196,13 @@ class TestWidgetCreation:
 class TestTaskListRefresh:
     """Tests for stale-row cleanup in TaskListWidget."""
 
-    def test_refresh_tasks_removes_stale_rows(self) -> None:
-        widget = TaskListWidget()
-        widget.on_mount()
+    @pytest.mark.asyncio
+    async def test_refresh_tasks_removes_stale_rows(self) -> None:
+        # DataTable.add_column measures its label against `self.app.console`,
+        # so `on_mount` raises LookupError on a widget that is not mounted in
+        # a running app. Drive it through `run_test` and let Textual dispatch
+        # Mount itself, which is also what exercises the real path.
+        app: App[object] = App()
         first = TaskRow(
             task_id="t-1",
             status="open",
@@ -218,11 +223,17 @@ class TestTaskListRefresh:
             elapsed="1s",
             session_id="s2",
         )
-        widget.refresh_tasks([first, second])
-        assert {str(key.value) for key in widget.rows} == {"t-1", "t-2"}
 
-        widget.refresh_tasks([first])
-        assert [str(key.value) for key in widget.rows] == ["t-1"]
+        async with app.run_test() as pilot:
+            widget = TaskListWidget()
+            await app.mount(widget)
+            await pilot.pause()
+
+            widget.refresh_tasks([first, second])
+            assert {str(key.value) for key in widget.rows} == {"t-1", "t-2"}
+
+            widget.refresh_tasks([first])
+            assert [str(key.value) for key in widget.rows] == ["t-1"]
 
 
 class TestAppInstantiation:

